crates/holochain/src/core/workflow/app_validation_workflow/unit_tests.rs

```diff
@@ -0,0 +1,506 @@
+use crate::{
+    conductor::space::TestSpace,
+    core::{
+        ribosome::{
+            guest_callback::validate::ValidateInvocation, real_ribosome::RealRibosome,
+            ZomesToInvoke,
+        },
+        workflow::app_validation_workflow::{
+            put_integrated, run_validation_callback_inner, Outcome,
+        },
+    },
+    sweettest::{SweetDnaFile, SweetInlineZomes},
+    test_utils::{test_network, test_network_with_events},
+};
+use fixt::fixt;
+use holo_hash::{hash_type::AnyDht, AgentPubKey, AnyDhtHash, DhtOpHash, HashableContentExtSync};
+use holochain_keystore::{test_keystore, MetaLairClient};
+use holochain_p2p::{event::HolochainP2pEvent, HolochainP2pDnaFixturator};
+use holochain_state::{host_fn_workspace::HostFnWorkspaceRead, mutations::insert_op};
+use holochain_types::{
+    dht_op::{DhtOp, DhtOpHashed, WireOps},
+    dna::DnaFile,
+    record::{SignedActionHashedExt, WireRecordOps},
+};
+use holochain_wasmer_host::module::ModuleCache;
+use holochain_zome_types::{
+    action::{ActionHashed, Dna},
+    chain::{ChainFilter, MustGetAgentActivityInput},
+    dependencies::holochain_integrity_types::{UnresolvedDependencies, ValidateCallbackResult},
+    entry::MustGetActionInput,
+    fixt::{ActionFixturator, CreateFixturator, DnaHashFixturator, SignatureFixturator},
+    judged::Judged,
+    op::{Op, RegisterAgentActivity},
+    record::SignedActionHashed,
+    timestamp::Timestamp,
+    validate::ValidationStatus,
+    zome::{IntegrityZomeDef, Zome},
+    Action,
+};
+use kitsune_p2p_types::ok_fut;
+use matches::assert_matches;
+use parking_lot::{Mutex, RwLock};
+use std::{
+    collections::HashSet,
+    sync::{
+        atomic::{AtomicI8, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+// test app validation with a must get action
+// where initially the action is not in the cache db
+// and is then added to it
+#[tokio::test(flavor = "multi_thread")]
+async fn validation_callback_must_get_action() {
+    let zomes =
+        SweetInlineZomes::new(vec![], 0).integrity_function("validate", move |api, op: Op| {
+            if let Op::RegisterAgentActivity(RegisterAgentActivity { action, .. }) = op {
+                if let Ok(_) = api.must_get_action(MustGetActionInput::new(action.to_hash())) {
```

Comment:
I can see how this works, but this is also a weird validation rule that nobody would actually write. It's postponing validation of the thing we already have in-hand, until we can get the same thing from somebody else. Maybe just make a note that it's weird and is this way because it makes the test simpler to write.